sources/appendix-value-extraction.asciidoc

```diff
@@ -0,0 +1,437 @@
+1. Sometimes constraints should apply to a value wrapped by a container type, but there is no type parameter to put the constraints to. JavaFX's `Property` hierarchy is the most prominent example: `@Email StringProperty emailProperty`. In such case the value should be implicitly extracted from the container if unambiguously doable.
+1. OPEN: Support for validation of nested collections: `List<Map<String, @NotNull String>> addressesByType`.
+1. Allow to configure type-level constraints and cascades via XML mappings
+1. Expose meta-data on type-level constraints and cascades in the constraint metadata API
+
+=== Non-requirements
+
+1. Constraints on type uses in local variables, invocations etc: `@NotNull String name = "Emmanuel";`, `new @NonEmpty @Readonly List<String>(myNonEmptyStringSet)`
+1. constraints on type uses in type definitions:
+
+[source,java]
+----
+class CustList extends List<@NotNull Customer> {
+}
+----
+
+=== Solution
+
+Value retrieval for cascaded validation and validation of type-level constraints is done via the `ValueExtractor` SPI:
+
+[source,java]
+----
+package javax.validation.valueretrieval;
+
+public interface ValueExtractor<T> {
+
+	void extractValues(T originalValue, ValueReceiver receiver);
+
+	interface ValueReceiver {
+
+		void value(Object object, String nodeName);
+
+		void iterableValue(Object object, String nodeName);
+
+		void indexedValue(Object object, String nodeName, int i);
+
+		void keyedValue(Object object, String nodeName, Object key);
+	}
+}
+----
+
+An extractor is tied to one specific type parameter of the type from which it extracts values. The `ExtractedValue` annotation is used to mark that type parameter. 
+
+As an example, this is how the implementation of the list extractor may look like:
+
+[source,java]
+----
+class ListValueExtractor implements ValueExtractor<List<@ExtractedValue ?>> {
+
+	@Override
+	public void extractValues(List<?> originalValue, ValueReceiver receiver) {
+		int i = 0;
+		for ( Object object : originalValue ) {
+			receiver.indexedValue( object, "<collection element>", i );
+			i++;
+		}
+	}
+}
+----
+
+The right callback methods (`indexedValue`, `keyedValue` etc.) must be called in order to allow for proper construction of the property path as per the rules layed out in the BV 1.1 spec.
+
+If a non-null value is passed for `nodeName`, a path node of type `TYPE_PARAMETER` will be appended to the property path. That's desirable for collection types for instance. If null is passed, no node will be appended, resulting in the same path as if the constraint had been given on the element itself instead of a type parameter. That's desirable for pure "wrapper types" such as `Optional`.
+
+In some cases, an extractor applies to the entire extracted type itself (e.g. for arrays). In this case the `@ExtractedValue` annotation is to be given on the extracted type itself:
```

Comment:
We should capture somewhere (here or in the wiki) why  `@ExtractedValue` lost the ability to refer to a supertype generic type. The main reason being that you can make an extractor for the supertype.

Note that it precludes:

*  a subtype to use better access methods for a faster container navigation (i.e. a better extractor implementation)
* JavaFX  `IntegerProperty` extractor to enforce `@ConstraintsApplyTo(WRAPPED_VALUE)` while JavaFX `Property<?>` extractor to enforce `@ConstraintsApplyTo(ANNOTATED_ELEMENT)`